File: app/js/src/gws/map/style.ts
import * as ol from 'openlayers';
import * as types from '../types';
import * as lib from '../lib';
import * as api from '../core/api';

export const DEFAULT_VALUES: types.Dict = {
    with_label: 'none',
    with_geometry: 'all',
    label_align: 'center',
    label_font_family: 'sans-serif',
    label_font_size: 12,
    label_font_style: 'normal',
    label_font_weight: 'normal',
    label_line_height: 1,
    label_max_scale: 1000000000,
    label_min_scale: 0,
    label_placement: 'middle',
    label_stroke_linejoin: 'round',
};


export class StyleManager implements types.IStyleManager {
    styles: { [cssSelector: string]: types.IStyle } = {};
    notFound: { [cssSelector: string]: boolean } = {};
    map: types.IMapManager;


    getFromSelector(cssSelector: string): types.IStyle | null {
        if (this.styles[cssSelector]) {
            return this.styles[cssSelector];
        }

        if (this.notFound[cssSelector]) {
            return null;
        }

        let [values, source] = getFromCss(cssSelector);
        if (values) {
            console.log('getFromSelector: found', cssSelector)
            let s = this.create(values, cssSelector);
            s.source = source;
            return s;
        }

        console.log('getFromSelector: NOT FOUND', cssSelector)
        this.notFound[cssSelector] = true;
        return null;
    }

    findFirst(selectors: Array<string>, geometryType: string = null, state: string = null) {
        let geom = geometryType ? '.' + geometryType.toLowerCase() : '';
        let spec = state ? '.' + state : '';

        for (let sel of selectors) {
            if (sel) {
                let style = (
                    this.getFromSelector(sel + geom + spec) ||
                    this.getFromSelector(sel + geom) ||
                    this.getFromSelector(sel + spec) ||
                    this.getFromSelector(sel)
                );
                if (style)
                    return style;
            }
        }
    }

    get names() {
        return Object.keys(this.styles)
    }

    at(cssSelector) {
        return this.styles[cssSelector] || null;
    }

    add(s) {
        this.styles[s.cssSelector] = s;
        return s;
    }

    get(arg: types.StyleArg) {
        if (!arg) {
            return null;
        }

        if (arg instanceof BaseStyle) {
            return arg;
        }

        if (typeof arg === 'string') {
            return this.getFromSelector(arg);
        }

        if (typeof arg === 'object') {
            let props = arg as api.core.StyleProps;

            if (props.cssSelector) {
                return this.getFromSelector(props.cssSelector);
            }

            if (props.values) {
                return this.create(props.values);
            }

            if (props.text) {

                let values = valuesFromCssText(props.text || '');

                if (values) {
                    return this.create(values);
                }
            }
        }

        console.warn('STYLE:invalid style', arg);
        return null;
    }

    whenStyleChanged(map: types.IMapManager, cssSelector?: string) {
        map.walk(map.root, layer => {
            let fs = (layer as types.IFeatureLayer).features;
            if (fs) {
                fs.forEach(f => f.redraw())
            }
        });
    }

    loadFromProps(props: api.core.StyleProps) {
        this.styles[props.cssSelector] = new Style(props.cssSelector, props.values);
        return this.styles[props.cssSelector];
    }

    get props(): Array<api.core.StyleProps> {
        let ps: Array<api.core.StyleProps> = [];

        for (let [_, style] of lib.entries(this.styles)) {
            ps.push(style.props);
        }

        return lib.compact(ps);
    }

    copy(style: types.IStyle, cssSelector: string = null) {
        return this.create(style.values, cssSelector)
    }

    protected create(values: types.Dict, cssSelector = null) {
        cssSelector = cssSelector || lib.uniqId('style');
        let s = new Style(cssSelector, values);
        this.styles[cssSelector] = s;
        return s;
    }

}

abstract class BaseStyle implements types.IStyle {
    cssSelector: string;
    values: types.Dict;
    source: string = '';

    get olFunction() {
        return (oFeature, resolution) => this.apply(oFeature.getGeometry(), oFeature.get('label'), resolution);
    }

    apply(geom, label, resolution) {
        return [];
    }

    get props() {
        return null;
    }

    update(values: types.Dict) {
    }
}


export class Style extends BaseStyle {

    protected cache: {
        markerImage?: ol.style.Image;
        pointImage?: ol.style.Image;
        shapeStyle?: ol.style.Style;
        iconImage?: ol.style.Icon;
        labelOptions?: types.Dict;
        labelMaxResolution?: number;
        labelMinResolution?: number;
        labelPlacement?: string;
    };

    constructor(cssSelector: string, values: types.Dict) {
        super();
        this.cssSelector = cssSelector;
        this.values = values;
    }

    get props() {
        return {
            cssSelector: this.cssSelector,
            values: this.values,
        }
    }

    update(values: types.Dict) {
        this.values = {...this.values, ...values};
        this.reset();
    }

    apply(geom, label, resolution) {
        if (!this.cache) {
            this.initCache();
        }

        let gt = geom.getType();
        let styles = [];

        if (this.cache.shapeStyle) {
            if (gt === 'Point' || gt === 'MultiPoint') {
                if (this.cache.pointImage) {
                    styles.push(new ol.style.Style({
                        image: this.cache.pointImage,
                        geometry: geom,
                    }))
                }
            } else {
                styles.push(this.cache.shapeStyle);
            }
        }

        if (this.cache.iconImage) {
            styles.push(new ol.style.Style({
                image: this.cache.iconImage,
                geometry: this.iconGeometry(geom),
            }));
        }


        if (this.cache.markerImage) {
            styles.push(new ol.style.Style({
                image: this.cache.markerImage,
                geometry: this.markerGeometry(geom),
            }));
        }

        let withLabel = label &&
            this.cache.labelOptions &&
            this.cache.labelMinResolution <= resolution &&
            resolution <= this.cache.labelMaxResolution;

        if (withLabel) {
            let opts = {...this.cache.labelOptions, text: label};
            let lp = this.cache.labelPlacement;

            if (gt === 'Point' || gt === 'MultiPoint' || gt === 'LineString' || gt === 'MultiLineString') {
                if (!('offsetY' in opts)) {
                    opts['offsetY'] = 20;
                }
                if (!lp) {
                    lp = 'end'
                }
            }

            let r = {
                text: new ol.style.Text(opts),
            };

            if (gt === 'Point' || gt === 'MultiPoint') {
                r['geometry'] = geom;
            } else if (gt === 'Circle') {
                r['geometry'] = new ol.geom.Point(geom.getCenter());
            } else if (lp === 'start') {
                r['geometry'] = new ol.geom.Point(geom.getFirstCoordinate());
            } else if (lp === 'end') {
                r['geometry'] = new ol.geom.Point(geom.getLastCoordinate());
            }

            styles.push(new ol.style.Style(r));
        }

        return styles;
    }

    protected initCache() {
        this.cache = {};

        let sv = {...DEFAULT_VALUES, ...this.values};

        if (sv.with_geometry === 'all') {

            let fill = olMakeFill(sv),
                stroke = olMakeStroke(sv);

            if (fill || stroke) {
                this.cache.shapeStyle = new ol.style.Style(compact({fill, stroke}));

                if (sv.point_size) {
                    this.cache.pointImage = new ol.style.Circle(compact({
                        fill, stroke, radius: this.values.point_size >> 1
                    }))
                }
            }

            this.cache.markerImage = olMakeMarker(sv);

            if (sv.icon) {
                this.cache.iconImage = new ol.style.Icon({
                    src: sv.icon,
                });
            }
        }

        if (sv.with_label === 'all') {

            this.cache.labelOptions = olLabelOptions(sv);

            if (this.cache.labelOptions) {

                this.cache.labelMinResolution = 0;
                this.cache.labelMaxResolution = 1e20;

                if ('label_min_scale' in sv)
                    this.cache.labelMinResolution = lib.scale2res(sv.label_min_scale);
                if ('label_max_scale' in sv)
                    this.cache.labelMaxResolution = lib.scale2res(sv.label_max_scale);

                this.cache.labelPlacement = sv.label_placement;
            }
        }
    }

    protected reset() {
        this.cache = null;
    }

    protected markerGeometry(geom) {
        let gt = geom.getType();

        if (gt === 'Point' || gt === 'Mutlipoint') {
            return geom;
        }

        let bounds = null

        if (gt === 'Polygon') {
            bounds = geom.getLinearRing(0).getCoordinates();
        } else if (gt === 'MultiPolygon') {
            bounds = geom.getPolygon(0).getLinearRing(0).getCoordinates();
        } else if (gt === 'Circle') {
            bounds = geom.getCenter();
        } else {
            bounds = geom.getCoordinates();
        }

        return new ol.geom.MultiPoint(bounds);
    }

    protected iconGeometry(geom) {
        let gt = geom.getType();

        if (gt === 'Point' || gt === 'Mutlipoint') {
            return geom;
        }
        if (gt === 'Polygon') {
            return (geom as ol.geom.Polygon).getInteriorPoint();
        }
        if (gt === 'MultiPolygon') {
            return (geom.getPolygon(0) as ol.geom.Polygon).getInteriorPoint();
        }

        let bounds = null;

        if (gt === 'Circle') {
            bounds = geom.getCenter();
        } else {
            bounds = geom.getCoordinates();
        }

        return new ol.geom.MultiPoint(bounds);
    }

}

export class CascadedStyle extends BaseStyle {
    styles: Array<types.IStyle>;

    constructor(cssSelector: string, styles: Array<types.IStyle>) {
        super();
        this.cssSelector = cssSelector;
        this.styles = styles;
    }

    apply(geom, label, resolution) {
        let res = [];

        this.styles.forEach(s => {
            res = res.concat(s.apply(geom, label, resolution));
        });

        return res;
    }
}

//

function olMakeFill(sv: types.Dict, prefix = '') {
    let r = compact({
        color: sv[prefix + 'fill'],
    });
    if (r && r.color)
        return new ol.style.Fill(r);
}

function olMakeStroke(sv: types.Dict, prefix = '') {
    let r = compact({
        color: sv[prefix + 'stroke'],
        lineDash: sv[prefix + 'stroke_dasharray'],
        lineDashOffset: sv[prefix + 'stroke_dashoffset'],
        lineCap: sv[prefix + 'stroke_linecap'],
        lineJoin: sv[prefix + 'stroke_linejoin'],
        miterLimit: sv[prefix + 'stroke_miterlimit'],
        width: sv[prefix + 'stroke_width'],
    });

    if (r && r.color)
        return new ol.style.Stroke(r);
}

function olMakeMarker(sv: types.Dict) {
    let marker = sv.marker || sv.marker_type,
        size = sv.marker_size;

    if (marker === 'circle' && size) {
        let r = compact({
            fill: olMakeFill(sv, 'marker_'),
            stroke: olMakeStroke(sv, 'marker_'),
            radius: size >> 1
        });
        if (r && (r.fill || r.stroke))
            return new ol.style.Circle(r);
    }
}

function olLabelOptions(sv: types.Dict) {
    let font = `
        ${sv.label_font_style} ${sv.label_font_weight} ${sv.label_font_size}px/${sv.label_line_height} ${sv.label_font_family}
    `;

    let r = compact({
        font: font.trim(),
        overflow: true,
        fill: olMakeFill(sv, 'label_'),
        stroke: olMakeStroke(sv, 'label_'),
        backgroundFill: !empty(sv.label_background) && new ol.style.Fill({color: sv.label_background}),
        padding: sv.label_padding,
        textAlign: sv.label_align,
        offsetX: sv.label_offset_x,
        offsetY: sv.label_offset_y,
    });

    if (r && (r.fill || r.stroke))
        return r;
}


//

function each(coll, fn) {
    for (let i = 0; i < coll.length; i++)
        fn(coll[i]);
}

function empty(x) {
    return typeof x === 'undefined' || x === null || x === '' || Number.isNaN(x);
}

function compact(obj): any {
    let d = {}, c = 0;

    each(Object.keys(obj), k => {
        if (!empty(obj[k])) {
            d[k] = obj[k];
            c++;
        }
    });

    return c ? d : null;
}


//

export function getFromCss(selector: string): [types.Dict, string] {


    function getStyleSheets() {
        try {
            // there can be stuff we're not allowed to read
            return document.styleSheets;
        } catch (e) {
            return [];
        }
    }

    function getRules(styleSheet) {
        try {
            return styleSheet.rules || styleSheet.cssRules || [];
        } catch (e) {
            return [];
        }
    }

    function matches(a, b) {
        return a && (a === b || String(a).endsWith(' ' + b));
    }

    let res = [];

    each(getStyleSheets(), s =>
        each(getRules(s), r => {
            if (matches(r.selectorText, selector) && r.style)
                res.push(r.style.cssText);
        })
    );

    let text = res.length ? res.pop() : null;

    if (text) {
        let values = valuesFromCssText(text);
        return [values, text];
    }

    return [null, ''];
}

// below is what we have in lib/style/parser.py


let _color_patterns = [
    /^#[0-9a-fA-F]{3}$/,
    /^#[0-9a-fA-F]{6}$/,
    /^#[0-9a-fA-F]{8}$/,
    /^rgb\(\d{1,3},\d{1,3},\d{1,3}\)$/,
    /^rgba\(\d{1,3},\d{1,3},\d{1,3},\d?(\.\d{1,3})?\)$/,
    /^[a-z]{3,50}$/,
];


let _color = (val) => {
    val = val.replace(/\s+/g, '');
    if (_color_patterns.some(p => val.match(p)))
        return val
};

let _px = (val) => {
    let m = val.match(/^(-?\d+)px/);
    return _int(m ? m[1] : val);
};

let _int = (val) => {
    val = Number(val);
    if (!Number.isNaN(val))
        return val;
};

let _intlist = (val) => {
    val = val.split(/[,\s]+/).map(_int);
    if (val.every(x => typeof x === 'number'))
        return val;
};

let _padding = (val) => {
    val = val.split(/[,\s]+/).map(_px);
    if (val.every(x => typeof x === 'number')) {
        if (val.length === 4)
            return val;
        if (val.length === 2)
            return [val[0], val[1], val[0], val[1]];
        if (val.length === 1)
            return [val[0], val[0], val[0], val[0]];
    }
};

let _enum = (cls) => {
    return (val) => {
        if (val in cls)
            return val
    }
};

let _str = (val) => {
    val = val.trim()
    return val || (void 0);
};


let _icon = (val) => {
    // remove url()
    let m = val.match(/^url\((.+?)\)$/);
    if (m) {
        val = m[1];
        if (val[0] === '"' || val[0] === "'") {
            val = val.slice(1, -1);
        }
    }
    return val;
};

let _Parser: any = {};

_Parser.fill = _color;
_Parser.stroke = _color;
_Parser.stroke_dasharray = _intlist;
_Parser.stroke_dashoffset = _px;
_Parser.stroke_linecap = _str;
_Parser.stroke_linejoin = _str;
_Parser.stroke_miterLimit = _px;
_Parser.stroke_width = _px;
_Parser.marker = _str;
_Parser.marker_type = _str;
_Parser.marker_fill = _color;
_Parser.marker_size = _px;
_Parser.marker_stroke = _color;
_Parser.marker_stroke_dasharray = _intlist;
_Parser.marker_stroke_dashoffset = _px;
_Parser.marker_stroke_linecap = _str;
_Parser.marker_stroke_linejoin = _str;
_Parser.marker_stroke_miterLimit = _px;
_Parser.marker_stroke_width = _px;
_Parser.with_geometry = _str;
_Parser.with_label = _str;
_Parser.label_align = _str;
_Parser.label_background = _color;
_Parser.label_fill = _color;
_Parser.label_font_family = _str;
_Parser.label_font_size = _px;
_Parser.label_font_style = _str;
_Parser.label_font_weight = _str;
_Parser.label_line_height = _int;
_Parser.label_max_scale = _int;
_Parser.label_min_scale = _int;
_Parser.label_offset_x = _px;
_Parser.label_offset_y = _px;
_Parser.label_padding = _padding;
_Parser.label_placement = _str;
_Parser.label_stroke = _color;
_Parser.label_stroke_dasharray = _intlist;
_Parser.label_stroke_dashoffset = _px;
_Parser.label_stroke_linecap = _str;
_Parser.label_stroke_linejoin = _str;
_Parser.label_stroke_miterLimit = _px;
_Parser.label_stroke_width = _px;
_Parser.point_size = _px;
_Parser.icon = _icon;
_Parser.offset_x = _px;
_Parser.offset_y = _px;


//

function valuesFromCssDict(d: types.Dict): types.Dict {
    let values = {};

    Object.keys(d).forEach(k => {
        let v = d[k];

        k = k.replace(/-/g, '_');
        if (k.startsWith('__'))
            k = k.slice(2);

        let fn = _Parser[k];
        if (fn) {
            v = fn(v);
            if (typeof v !== 'undefined')
                values[k] = v;
        }
    });

    return values as types.Dict;
}

export function valuesFromCssText(text: string): types.Dict {
    let d = {};

    each((text || '').split(';'), r => {
        r = r.trim();
        if (!r)
            return;
        let m = r.match(/^(.+?):(.+)/);
        d[m[1].trim()] = m[2].trim();
    });

    return valuesFromCssDict(d);
}

File: app/gws/lib/style.py
import base64
import re

import gws
import gws.types as t
import gws.lib.net


class Type(t.Enum):
    css = 'css'
    cssSelector = 'cssSelector'


class Values(gws.Data):
    fill: gws.Color

    stroke: gws.Color
    stroke_dasharray: t.List[int]
    stroke_dashoffset: int
    stroke_linecap: t.Literal['butt', 'round', 'square']
    stroke_linejoin: t.Literal['bevel', 'round', 'miter']
    stroke_miterlimit: int
    stroke_width: int

    marker: t.Literal['circle', 'square', 'arrow', 'cross']
    marker_fill: gws.Color
    marker_size: int
    marker_stroke: gws.Color
    marker_stroke_dasharray: t.List[int]
    marker_stroke_dashoffset: int
    marker_stroke_linecap: t.Literal['butt', 'round', 'square']
    marker_stroke_linejoin: t.Literal['bevel', 'round', 'miter']
    marker_stroke_miterlimit: int
    marker_stroke_width: int

    with_geometry: t.Literal['all', 'none']
    with_label: t.Literal['all', 'none']

    label_align: t.Literal['left', 'right', 'center']
    label_background: gws.Color
    label_fill: gws.Color
    label_font_family: str
    label_font_size: int
    label_font_style: t.Literal['normal', 'italic']
    label_font_weight: t.Literal['normal', 'bold']
    label_line_height: int
    label_max_scale: int
    label_min_scale: int
    label_offset_x: int
    label_offset_y: int
    label_padding: t.List[int]
    label_placement: t.Literal['start', 'end', 'middle']
    label_stroke: gws.Color
    label_stroke_dasharray: t.List[int]
    label_stroke_dashoffset: int
    label_stroke_linecap: t.Literal['butt', 'round', 'square']
    label_stroke_linejoin: t.Literal['bevel', 'round', 'miter']
    label_stroke_miterlimit: int
    label_stroke_width: int

    point_size: int
    icon: str

    offset_x: int
    offset_y: int


class Data(gws.StyleData):
    name: str
    type: Type
    text: str
    values: Values


class Config(gws.Config):
    """Feature style"""

    type: Type  #: style type
    name: t.Optional[str]  #: style name
    text: t.Optional[str]  #: raw style content
    values: t.Optional[dict]  #: style values


class Props(gws.Props):
    type: Type
    values: dict
    text: str = ''
    name: str = ''


class Object(gws.Node, gws.IStyle):
    data: gws.StyleData

    @property
    def props(self):
        d = t.cast(Data, self.data)
        return gws.Props(
            type=d.type,
            values=vars(d.values),
            text=d.text or '',
            name=d.name or '')


def from_config(cfg: Config) -> Object:
    return from_props(t.cast(gws.Props, cfg))


def from_props(p: gws.Props) -> Object:
    typ = p.get('type', 'css')
    data = gws.StyleData(p, type=typ)

    if typ == 'css':
        val = p.get('values')
        if val:
            data.set('values', values_from_dict(gws.as_dict(val)))
        else:
            data.set('values', values_from_text(p.get('text')))

    obj = Object()
    obj.data = data
    return obj


def values_from_dict(d: dict) -> Values:
    values = Values(_DEFAULT_VALUES)

    for k, v in d.items():
        if v is None:
            continue
        k = k.replace('-', '_')
        if k.startswith('__'):
            k = k[2:]
        fn = gws.get(_Parser, k)
        if fn:
            v = fn(v)
            if v is not None:
                setattr(values, k, v)

    return values


# @TODO use a real parser
def values_from_text(text) -> Values:
    d = {}
    for r in text.split(';'):
        r = r.strip()
        if not r:
            continue
        r = r.split(':')
        d[r[0].strip()] = r[1].strip()
    return values_from_dict(d)


def parse_icon(val):
    return _icon(val)


_DEFAULT_VALUES = gws.Data(
    fill=None,

    stroke=None,
    stroke_dasharray=[],
    stroke_dashoffset=0,
    stroke_linecap='butt',
    stroke_linejoin='miter',
    stroke_miterlimit=0,
    stroke_width=0,

    marker_size=0,
    marker_stroke_dasharray=[],
    marker_stroke_dashoffset=0,
    marker_stroke_linecap='butt',
    marker_stroke_linejoin='miter',
    marker_stroke_miterlimit=0,
    marker_stroke_width=0,

    with_geometry='all',
    with_label='all',

    label_align='center',
    label_font_family='sans-serif',
    label_font_size=12,
    label_font_style='normal',
    label_font_weight='normal',
    label_line_height=1,
    label_max_scale=1000000000,
    label_min_scale=0,
    label_offset_x=0,
    label_offset_y=0,
    label_placement='middle',
    label_stroke_dasharray=[],
    label_stroke_dashoffset=0,
    label_stroke_linecap='butt',
    label_stroke_linejoin='miter',
    label_stroke_miterlimit=0,
    label_stroke_width=0,

    point_size=10,
    icon='',

    offset_x=0,
    offset_y=0,
)

_color_patterns = (
    r'^#[0-9a-fA-F]{3}$',
    r'^#[0-9a-fA-F]{6}$',
    r'^#[0-9a-fA-F]{8}$',
    r'^rgb\(\d{1,3},\d{1,3},\d{1,3}\)$',
    r'^rgba\(\d{1,3},\d{1,3},\d{1,3},\d?(\.\d{1,3})?\)$',
    r'^[a-z]{3,50}$',
)


def _color(val):
    val = re.sub(r'\s+', '', str(val))
    if any(re.match(p, val) for p in _color_patterns):
        return val


def _icon(val):
    if not val:
        return
    s = val
    m = re.match(r'^url\((.+?)\)$', val)
    if m:
        s = m.group(1).strip('\'\"')
    try:
        return _to_data_url(s)
    except Exception as e:
        raise gws.Error(f'cannot load {val!r}') from e


def _to_data_url(val):
    if val.startswith('data:'):
        return val
    # @TODO security, this should be only allowed in a trusted context
    if re.match(r'^https?:', val):
        svg = gws.lib.net.http_request(val).content
    else:
        svg = gws.read_file_b(val)
    return 'data:image/svg+xml;base64,' + base64.standard_b64encode(svg).decode('utf8')


def _px(val):
    if isinstance(val, int):
        return val
    m = re.match(r'^(-?\d+)px', str(val))
    return _int(m.group(1) if m else val)


def _int(val):
    if isinstance(val, int):
        return val
    try:
        return int(val)
    except:
        pass


def _intlist(val):
    if not isinstance(val, (list, tuple)):
        val = re.split(r'[,\s]+', str(val))
    val = [_int(x) for x in val]
    if any(x is None for x in val):
        return None
    return val


def _padding(val):
    if not isinstance(val, (list, tuple)):
        val = re.split(r'[,\s]+', str(val))
    val = [_px(x) for x in val]
    if any(x is None for x in val):
        return None
    if len(val) == 4:
        return val
    if len(val) == 2:
        return [val[0], val[1], val[0], val[1]]
    if len(val) == 1:
        return [val[0], val[0], val[0], val[0]]
    return None


def _enum(cls):
    def _check(val):
        if val in vars(cls):
            return val

    return _check


def _str(val):
    val = str(val).strip()
    return val or None


class _Parser:
    fill = _color

    stroke = _color
    stroke_dasharray = _intlist
    stroke_dashoffset = _px
    stroke_linecap = _enum('stroke_linecap')
    stroke_linejoin = _enum('stroke_linejoin')
    stroke_miterlimit = _px
    stroke_width = _px

    marker = _enum('marker')
    marker_fill = _color
    marker_size = _px
    marker_stroke = _color
    marker_stroke_dasharray = _intlist
    marker_stroke_dashoffset = _px
    marker_stroke_linecap = _enum('stroke_linecap')
    marker_stroke_linejoin = _enum('stroke_linejoin')
    marker_stroke_miterlimit = _px
    marker_stroke_width = _px

    with_geometry = _enum('x')
    with_label = _enum('x')

    label_align = _enum('x')
    label_background = _color
    label_fill = _color
    label_font_family = _str
    label_font_size = _px
    label_font_style = _enum('x')
    label_font_weight = _enum('x')
    label_line_height = _int
    label_max_scale = _int
    label_min_scale = _int
    label_offset_x = _px
    label_offset_y = _px
    label_padding = _padding
    label_placement = _enum('x')
    label_stroke = _color
    label_stroke_dasharray = _intlist
    label_stroke_dashoffset = _px
    label_stroke_linecap = _enum('x')
    label_stroke_linejoin = _enum('x')
    label_stroke_miterlimit = _px
    label_stroke_width = _px

    point_size = _px
    icon = _icon

    offset_x = _px
    offset_y = _px
